lib/CGI.pm

```diff
@@ -622,16 +622,8 @@ sub init {
 	  last METHOD;
       }
 
-      # If method is GET, HEAD or DELETE, fetch the query from
-      # the environment.
-      if ($is_xforms || $meth=~/^(GET|HEAD|DELETE)$/) {
-          $query_string = $self->_get_query_string_from_env;
-         $self->param($meth . 'DATA', $self->param('XForms:Model'))
-             if $is_xforms;
-	      last METHOD;
-      }
-
-      if ($meth eq 'POST' || $meth eq 'PUT') {
+      # First, let's try to get data from request body
+      if ($meth=~/^(GET|HEAD|DELETE|POST|PUT)$/) {
```

Comment:
Unfortunately trying to be generic is likely to break backwards compatibility for some users.